sets/slices.go

```diff
@@ -0,0 +1,196 @@
+func IntersectStable[T comparable](slices ...[]T) []T {
+	if len(slices) == 0 {
+		return []T{}
+	}
+
+	// Make a copy of the first slice; we're going to gradually remove elements.
+	final := append([]T{}, slices[0]...)
+
+	// Remove duplicates
+	finalSeen := make(map[T]struct{}, len(final))
+	var i int
+	for _, v := range final {
+		if _, ok := finalSeen[v]; !ok {
+			finalSeen[v] = struct{}{}
+			final[i] = v
+			i++
+		}
+	}
+	for j := i; j < len(final); j++ {
+		final[j] = *new(T)
+	}
+	final = final[:i:len(final)]
+
+	for _, s := range slices[1:] {
+		// Short-circuit: if our intersection is already the empty set, we can
+		// return now.
+		if len(final) == 0 {
+			return []T{}
+		}
+
+		var i int
+		for _, v := range final {
+			if sliceContains(s, v) {
+				final[i] = v
+				i++
+			}
+		}
+		// Delete unused elements, nil out unused space so we don't leak.
+		for j := i; j < len(final); j++ {
+			final[j] = *new(T)
+		}
+		final = final[:i:len(final)]
+	}
+	return final
+}
+
+// Union finds the union of all slices, where union is defined as the
+// combination of all elements in all slices. The elements are returned in the
+// order in which they first appear in the slices. Duplicate elements are
+// removed. This function always returns an allocated slice, even if the union
+// is the empty set.
+//
+// It does not modify any of the given slices, but also does not deep copy any
+// values. That means the returned slice may have elements that point to the
+// same objects as in the original slice.
+func Union[T comparable](slices ...[]T) []T {
+	if len(slices) == 0 {
+		return []T{}
+	}
+
+	// Pre-compute the maximum possible allocation to minimize allocs.
+	var alloc int
+	for _, s := range slices {
+		alloc += len(s)
+	}
+	final := make([]T, 0, alloc)
+	seen := make(map[T]struct{}, alloc)
+
+	// Iterate over all maps and combine elements.
+	for _, s := range slices {
+		for _, v := range s {
+			if _, ok := seen[v]; ok {
+				// We already saw this element
+				continue
+			}
+
+			final = append(final, v)
+			seen[v] = struct{}{}
+		}
+	}
+	return final[:len(final):len(final)]
+}
+
+// Subtract finds the differece (subtraction), where difference is defined as
+// the removal of all elements from s0 that exist in sn. The elements are
+// returned in the order in which they appeared in s0. This function always
+// returns an allocated slice, even if the difference is the empty set.
+//
+// All duplicate elements that match a subtraction are removed. Duplicate
+// elements that do not match a subtraction are preserved.
+//
+// It does not modify any of the given slices, but also does not deep copy any
+// values. That means the returned slice may have elements that point to the
+// same objects as in the original slice.
+func Subtract[T comparable](slices ...[]T) []T {
+	if len(slices) == 0 {
+		return []T{}
+	}
+
+	// Make a copy of the first slice; we're going to gradually remove elements.
+	final := append([]T{}, slices[0]...)
```

Comment:
Would it be faster/cleaner to not do the first call to append?

First generate the toDelete map from `slices[1:]` and then iterate once through `slices[0]` appending to `final` only if `slices[0][i]` is not in `toDelete`

Reply:
Maybe rename `toDelete` to `toSkip`, `toSubtract` or `toOmit`